### PR TITLE
issue-255, $size operator

### PR DESCRIPTION
**Pull request by taaraora**
`$size` query comparison operator for array was implemented.

### CODE REVIEW

File: internal/handlers/handler_test.go [hunk 1]
```diff
@@ -80,6 +80,90 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
 	return actual
 }
 
+func TestFindArray(t *testing.T) {
+	t.Parallel()
+	ctx, handler, pool := setup(t, nil)
+	schema := testutil.Schema(ctx, t, pool)
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:86]: This function creates a test-specific PostgreSQL schema / MongoDB database and drops it if the test passes. There is no need to do the manual cleanup with `cleanArrayTestData`.

File: internal/handlers/handler_test.go [hunk 2]
```diff
@@ -80,6 +80,90 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
...
+	loadArrayTestData(ctx, t, handler, schema)
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:87]: Let's start making a test dataset that would be shared between various query tests. I was thinking about documents with two fields: string `_id` for identification and `value` of various types. That would allow us to test edge cases like calling `$size` operators on NaN double value. https://github.com/FerretDB/dance/pull/33/files#diff-ee104391e4d88f69dce464c4f136bdeeb5c3d9a3b494a639416e6e0454afd4c2R86-R157

File: internal/handlers/handler_test.go [hunk 3]
```diff
@@ -80,6 +80,90 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
...
+	defer cleanArrayTestData(ctx, t, handler, schema)
+
+	req := must.NotFail(types.NewDocument(
+		"find", "dmts",
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:91]: `testutil.TableName` instead of dmts

File: internal/handlers/jsonb1/where.go [hunk 4]
```diff
@@ -142,6 +142,13 @@ func fieldExpr(field string, expr *types.Document, p *pg.Placeholder) (sql strin
 			// {field: {$gte: value}}
 			sql += "_jsonb->" + p.Next() + " >="
 			argSql, arg, err = scalar(value, p)
+		case "$size":
+			// { field: { $size: 2 } }
+			sql += "jsonb_array_length(_jsonb::jsonb->" + p.Next() + ") = "
+
+			var sizeValue any
+			sizeValue, err = expr.Get(expr.Command())
+			argSql, arg = fmt.Sprintf("%s", p.Next()), []any{sizeValue}
```
Comment (AlekSi) [inline internal/handlers/jsonb1/where.go:151]:
```suggestion
			argSql, arg = p.Next(), []any{sizeValue}
```

File: internal/handlers/jsonb1/where.go [hunk 5]
```diff
@@ -142,6 +142,13 @@ func fieldExpr(field string, expr *types.Document, p *pg.Placeholder) (sql strin
 			// {field: {$gte: value}}
 			sql += "_jsonb->" + p.Next() + " >="
 			argSql, arg, err = scalar(value, p)
+		case "$size":
+			// { field: { $size: 2 } }
+			sql += "jsonb_array_length(_jsonb::jsonb->" + p.Next() + ") = "
```
Comment (AlekSi) [inline internal/handlers/jsonb1/where.go:147]: I _think_ `::jsonb` casting should not be required there – `_jsonb` column already has this type.

File: internal/handlers/jsonb1/where.go [hunk 6]
```diff
@@ -142,6 +142,13 @@ func fieldExpr(field string, expr *types.Document, p *pg.Placeholder) (sql strin
...
+
+			var sizeValue any
+			sizeValue, err = expr.Get(expr.Command())
```
Comment (AlekSi) [inline internal/handlers/jsonb1/where.go:150]:
```suggestion
			sizeValue, err = expr.Get(op)
```

But we should also check type

File: internal/handlers/jsonb1/where.go (hunk 5 above)
Comment (taaraora) [reply internal/handlers/jsonb1/where.go]: I totally agree.

File: internal/handlers/handler_test.go (hunk 1 above)
Comment (taaraora) [reply internal/handlers/handler_test.go:87]: Got it, `cleanArrayTestData` is removed, thank you.

File: internal/handlers/handler_test.go (hunk 2 above)
Comment (taaraora) [reply internal/handlers/handler_test.go]: I tried to mimic, that helped to catch new a bug.
As I got from documentation `$size` is used only with arrays.

File: internal/handlers/handler_test.go (hunk 3 above)
Comment (taaraora) [reply internal/handlers/handler_test.go]: Thank you, fixed.

File: internal/handlers/jsonb1/where.go (hunk 4 above)
Comment (taaraora) [reply internal/handlers/jsonb1/where.go]: Thank you, fixed.

File: internal/handlers/jsonb1/where.go (hunk 6 above)
Comment (taaraora) [reply internal/handlers/jsonb1/where.go]: I've updated the SQL query with type check, thank you!

Comment (AlekSi) [reply internal/handlers/jsonb1/where.go]: We should check that the `$size` operator is called on the array with integer (btw, can it be float?) argument, and return MongoDB-compatible errors if types are different. We should also have tests for those errors.

I think that invoking `jsonb_array_length` on the argument and checking the error code later might be easier than doing a type checking in SQL.

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/296#discussion_r797619467

File: internal/handlers/jsonb1/where.go (hunk 6 above)
Comment (taaraora) [reply internal/handlers/jsonb1/where.go]: 1. I've added tests for `$size`.
2. Regarding invoking `jsonb_typeof()=array` in SQL. When I filled up the collection with mixed types values `jsonb_array_length()` function just started throwing exceptions for non-array values. At the same time, Mongo handles such cases silently skipping non-array values. So I think that type check is justified.

**Review by AlekSi — COMMENTED** [review]
I added integration tests. Please make this one pass: https://github.com/FerretDB/dance/blob/51b6c9ba578243af1763d457d255161a151909ad/tests/ferret/core_test.go#L193-L200

Please also merge with the `main` branch

File: internal/handlers/jsonb1/where.go (hunk 6 above)
Comment (AlekSi) [reply internal/handlers/jsonb1/where.go]: > silently skipping non-array values

That what I was missing. Thanks.

File: internal/handlers/common/validate.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+package common
+
+import (
+	"fmt"
+	"math"
+)
+
+func IsPositiveInteger(value any) error {
```
Comment (AlekSi) [inline internal/handlers/common/validate.go:22]: Let's not generalize and move to common that function just yet. Let's keep it separate, but in jsonb1 package, with a name something like `validateSizeArg`. Once we have several similar fucntions, then we would generalize them.

I think that function can also use `common.NewError` instead of error strings

File: internal/handlers/handler_test.go [hunk 8]
```diff
@@ -37,6 +40,12 @@ import (
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
+type testCase struct {
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:43]: That's a common type for query tests, but not for inserts or updates, for example. Let's move it inside TestSize

File: internal/handlers/handler_test.go [hunk 9]
```diff
@@ -69,15 +87,197 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
...
+
+		t.Run(tcName, func(t *testing.T) {
+			t.Parallel()
+
+			actual, err := handleWithError(ctx, t, handler, tc.req)
+			expected := must.NotFail(types.NewDocument(
+				"cursor", must.NotFail(types.NewDocument(
+					"firstBatch", tc.resp,
+					"id", int64(0),
+					"ns", schema+"."+testutil.TableName(parentT),
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:219]: Maybe we can make `ns` before subtest and avoid using `parentT`?

File: internal/handlers/handler_test.go [hunk 10]
```diff
@@ -69,15 +87,197 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
...
+				)),
+				"ok", float64(1),
+			))
+
+			if tc.err != nil && assert.ErrorAs(t, tc.err, &err) {
+				assert.Equal(t, tc.err.Error(), err.Error())
+				// TODO: compare error code
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:226]: That's actually quite important. `assert.Equal(t, tc.err, err)` does not work?

File: internal/handlers/handler_test.go [hunk 11]
```diff
@@ -69,15 +87,197 @@ func handle(ctx context.Context, t *testing.T, handler *Handler, req *types.Docu
...
+			} else {
```
Comment (AlekSi) [inline internal/handlers/handler_test.go:227]:
```go
if tc.err != nil {
  // requires and asserts
  return
}
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: internal/handlers/jsonb1/where.go [hunk 12]
```diff
@@ -142,6 +142,17 @@ func fieldExpr(field string, expr *types.Document, p *pg.Placeholder) (sql strin
...
+		case "$size":
+			// { field: { $size: 2 } }
+			var sizeValue any
+			sizeValue, err = expr.Get(op)
+			if err = common.IsPositiveInteger(sizeValue); err != nil {
+				err = common.NewError(common.ErrBadValue, err)
+				return
+			}
+
+			sql += "jsonb_typeof(_jsonb->" + p.Next() + ") = 'array' AND jsonb_array_length(_jsonb->" + p.Next() + ") = "
```
Comment (AlekSi) [inline internal/handlers/jsonb1/where.go:154]: That should be the same placeholder (number) to avoid sending the same argument (field name) twice

**Review by AlekSi — APPROVED** [review]
